R/n_imaging.r

```diff
@@ -41,51 +39,83 @@
-#' imaging test (usually `"performed_date_time"`).
-#' @param mapvar (`character`)
-#' Name of the column in `imaging` table containing the imaging test type
-#' (usually `"modality_mapped"`).
+#'
+#' @param exclude_ed (`logical`)
+#' Whether to exclude tests in emergency department. When set to `TRUE`, tests performed
+#' in ED (testing time before admission time) will not be counted. When set to `FALSE`,
+#' tests will not be filtered by time and all tests in radiology table will be counted.
+#' Please be aware that this may include tests before triage time, tests after discharge
+#' time, and tests without testing time.
+#'
+#' Tests in ED are defined as `ordered_date_time` earlier than `admission_date_time`.
+#' When `ordered_date_time` is not available, `performed_date_time` is used instead.
 #'
 #' @return
 #' data.table with the same number of rows as input "cohort", with additional
 #' derived numeric fields labelled as "n_img_xray_derived",  "n_img_ct_derived",
 #' "n_img_mri_derived",  "n_img_us_derived", "n_img_other_derived",
 #' "n_img_int_derived" and "n_img_ct_mri_us_derived".
 #'
+#' @note
+#' Currently, the function does not take data availability into account. For
+#' patients without imaging tests, the function will return 0 in result columns.
+#' User should check radiology data availability and decide whether the imputed
+#' `0`s are appropriate or not.
+#'
 #' @export
-n_imaging <- function(cohort,
-                      imaging,
-                      dtvar = "performed_date_time",
-                      mapvar = "modality_mapped") {
-
-
+n_imaging <- function(dbcon,
+                      cohort,
+                      exclude_ed = FALSE) {
+  # warning messages
   mapping_message("imaging modalities")
+  availability_message("radiology")
 
-  ## remap variable names in case field names change in the database
+  # check input type and column name
+  check_input(cohort, argtype = c("data.table", "data.frame"), colnames =  c("genc_id"))
+  check_input(exclude_ed, arginput = "logical")
   cohort <- coerce_to_datatable(cohort)
-  imaging <- coerce_to_datatable(imaging)
 
-  res <- cohort[, .(genc_id)]
+  # identify tables in db, currently the function does not work on radiology so
+  # it is hard coded as "radiology" in query
+  admdad_table <- find_db_tablename(dbcon, "admdad", verbose = FALSE)
+
+  # query db to pull imaging data
+  imaging <- dbGetQuery(
+    dbcon,
+    paste0(
+      ifelse(exclude_ed == TRUE,
+            # filter by admission date time and exclude tests before admission
+            paste0("with temp as (
+              select r.*,a.admission_date_time,
+              case when r.ordered_date_time is null or r.ordered_date_time = '' or
+              r.ordered_date_time = ' ' then r.performed_date_time >= a.admission_date_time
+              else r.ordered_date_time >= a.admission_date_time end as case_result
+              from radiology r
```

Comment:
Thanks for catching this Anne and yes I hardcoded the table name here because `fin_db_tablename` does not allow `radiology` to be the input table name yet. You plan sounds great. Thanks!

Reply:
Perfect, I just pushed a commit that adds the radiology table to `find_db_tablename` so this should hopefully work now.